### DISCUSSION TITLE
Feature request - composer outdated with pinned/locked versions

### DISCUSSION THREAD

**Discussion by sebb-m0** [composer/composer#11577] [General]
We are growing our PHP projects and pipelines and want to get information, when a version of a package is pinned as a direct dependency like:

```json
    "phpunit": "10.2.1",
```

So if this is "outdated" and pinned here as a hard version constraint, I would like to be informed too.

## example

```bash
composer outdated --pinned
```

so that i can loop over the projects and get an information, if there are any hard pinned versions and what is the last possible version to update.

sometimes we need to pin versions like this, to be compatible with 3rd party libraries, which can't handle the latest version.

(our example: codeception 5.x can't handle phpunit 10.3)

**Comment by Seldaek** [comment]
Not sure why the fact they are pinned matters.

If I require for example:

```
        "monolog/monolog": "2.3.5"
```

Then outdated gives me the info that there is a new version:

With --minor-only:

```
monolog/monolog 2.3.5 2.9.1 Sends your logs to files, sockets, inboxes, databases and various web services
````

And without:

```
monolog/monolog 2.3.5 3.4.0 Sends your logs to files, sockets, inboxes, databases and various web services
```

Besides, I would recommend pinning with `~` or `*` like `~10.2.1` in your example would allow updating to 10.2.2, 10.2.3 etc but not 10.3. That lets you receive patch releases.

**Reply by sebb-m0** [reply]
Sure. But this is not what I want to achieve.

The thing is, that I need the information about these hard pinned versions. The reason, why they pinned are almost third party libraries, which are patching versions, but introduce breaking changes or incompatibilities with other third party libraries. These pinned versions should be a work around, which have to be solved a short period later, when the breaking change/incompatibility is fixed somehow.

In that regard, I just wrote myself a small shell script with grep, which finds hard pinned versions in the composer.json. I just miss the functionality in composer directly, but maybe it's too specific. I thought maybe some else could take advantage from this feature as well.

```bash
#!/usr/bin/env bash

if [ $# -ne 1 ]; then
    echo "Usage: $0 <path-to-json-file>"
    exit 1
fi

path_to_json_file=$1

# regex which finds pinned versions, 1.2.3 is an example of a pinned version
regex_pattern='"[0-9]+\.[0-9]+\.[0-9]+'

strings_to_ignore=(
  "composer/package-versions-deprecated"
)

echo "Checking for pinned versions in $path_to_json_file"
result=$(grep -E "$regex_pattern" "$path_to_json_file" | grep -v -F "${strings_to_ignore[@]}")

if [ -n "$result" ]; then
    echo "Found pinned versions:"
    echo "$result"
    exit 1
fi

echo "No pinned versions found."
exit 0
```

**Reply by Seldaek** [reply]
Yeah I see but it doesn't look to me like something that belongs in `outdated` tbh, it's more a reporting thing for pinned versions.. Note that running `composer validate` will warn about these, for example:

```
- require.symfony/security : exact version constraints (4.0.5) should be avoided if the package follows semantic versioning
```

**Comment by sebb-m0** [comment]
Oh, I wasn't aware of that. Maybe this will help a lot. 👍
